File: MultiThreading/Node.cpp
#include "Node.h"
#include "Search.h"

// 构造函数
Node::Node(float priorProb, float cPuct, Node* parent)
{
	this->Q = 0.0f;
	//this->U = 0.0;
	this->N = 0;
	this->P = priorProb;
	this->cPuct = cPuct;
	this->parent = parent;
}

// 析构函数
Node::~Node()
{
	for (auto it : children)
		delete it.second;
}

// 选择得分最高的节点
// 返回std::pair<int, Node*>(maxAction, maxNode)
std::pair<int, Node*> Node::select()
{
	std::shared_lock<std::shared_mutex> lk(nodeLock);

	int maxAction = -1;
	float score = -10000;
	Node* maxNode = nullptr;

	for (std::map<int, Node*>::const_iterator it = children.begin(); it != children.end(); it++)
	{
		float newScore = it->second->getScore();
		if (newScore > score)
		{
			maxAction = it->first;
			maxNode = it->second;
			score = newScore;
		}
	}
	// 增加虚拟损失
	if (maxNode)
		maxNode->virtualLoss++;

	//for (std::map<int, Node*>::iterator it = children.begin();
	//	it != children.end(); it++)
	//{
	//	if (it->first == maxAction)
	//	{
	//		it->second->virtualLoss++;
	//	}
	//}
	return std::pair<int, Node*>(maxAction, maxNode);
}

// 扩展节点
// 传入std::map<int, double>
void Node::expand(const std::map<int, float>& actionProb)
{
	// 外部加锁
	if (children.empty())
	{
		for (std::map<int, float>::const_iterator it = actionProb.begin();
			it != actionProb.end(); it++)
		{
			Node* childNode = new Node(it->second, cPuct, this);
			children.insert(std::pair<int, Node*>(it->first, childNode));
		}
	}
}

// 更新 Q和N
void Node::update(float value)
{
	// 加自旋锁操作，短期操作，不sleep
	std::lock_guard<SpinLock> lk(spinLock);

	virtualLoss--;
	Q = (N * Q + value) / (N + 1);
	N++;
}

// 反向传播
void Node::backup(float value)
{
	if (parent)
	{
		parent->backup(-value);
	}
	update(value);
}

// 获取得分
// 返回score
float Node::getScore()
{
	std::lock_guard<SpinLock> lk(spinLock);

	if (parent == nullptr)
		return 0;

	float U = cPuct * P * sqrtf(parent->N) / (N + 1.0f + virtualLoss * cVirtualLoss);

	float QQ = N ? Q : -(P * 2.0f - 1.0f); //std::max(-1.0f, std::min(1.0f, -parent->Q + cVirtualQDelta));
	int NN = std::max<int>(N, 1);
	float virtualQ = (QQ * NN + std::max(QQ, -parent->Q) * virtualLoss * cVirtualLoss) / (NN + virtualLoss * cVirtualLoss);
	// 注意 Q是对手胜率
	return U - virtualQ;
}

float Node::getUpQ() const
{
	return (-Q * N + (P * 2.0f - 1.0f)) / (N + 1);
}

// 是否为叶子节点
bool Node::isLeafNode(bool lock)
{
	if (lock) nodeLock.lock_shared();
	bool flag = children.empty();
	if (lock) nodeLock.unlock_shared();
	return flag;
}

File: SingleThreading/node.py
from math import sqrt
from typing import Tuple, Iterable

class Node:
    """
    蒙特卡洛树节点。
    """

    def __init__(self, prior_prob: float, c_puct: float = 1, parent=None):
        """
        初始化节点。
        :param prior_prob: float，先验概率P(s,a)
        :param c_puct: float，探索系数
        :param parent: None，父节点
        """
        self.Q = 0
        self.U = 0
        self.N = 0
        self.P = prior_prob
        self.c_puct = c_puct
        self.parent = parent
        # {action: Node}
        self.children = {}  # type -> Dict[int, Node]

    def select(self) -> tuple:
        """
        选择score最大的子节点和该节点对应的action
        :return: tuple[int, None] action: int, child: None
        """
        for k,v in self.children.items():
            if(v.N==0):#没计算过神经网络
                v.Q = -v.parent.Q + 0.1

        action, node = max(self.children.items(), key=lambda item: item[1].get_score())
        return action, node

    def expand(self, action_prob: Iterable[Tuple[int, float]]):
        """
        拓展节点。
        :param action_prob: Iterable
        每个元素都是[action, prior_probd]的元组，根据这个远组创建子节点
        action_prob的长度为当前棋盘的可用落点的总数
        :return:
        """
        #self.N = 1
        #self.Q = value


        for action, prior_prob in action_prob:
            self.children[action] = Node(prior_prob, self.c_puct, self)


    def __update(self, value: float):
        """
        更新节点的访问次数 N(s,a) 和节点的累计平均奖励Q(s,a)
        :param value:float -> 用来更新节点的内部数据。
        :return:
        """
        self.Q = (self.N * self.Q + value) / (self.N + 1)
        self.N += 1

    def backup(self, value: float):
        """
        反向传播。
        :param value:
        :return:
        """
        if self.parent:
            self.parent.backup(-value)
        self.__update(value)

    def get_score(self):
        """
        计算节点得分。
        :return:
        """
        #print( self.N,self.P,self.Q,self.parent.N)
        self.U = self.c_puct * self.P * sqrt(self.parent.N) / (1 + self.N)
        score = self.U - self.Q #注意self.Q是对手胜率
        return score

    def is_leaf_node(self):
        """
        是否为叶子节点。
        :return:
        """
        return len(self.children) == 0
